File: use-cases/code-algorithms/python/task_list_merge.py
import copy

def merge_task_lists(local_tasks, remote_tasks):
    """
    Merge two task lists with conflict resolution.

    Args:
        local_tasks: Dictionary of tasks from local source {task_id: task}
        remote_tasks: Dictionary of tasks from remote source {task_id: task}

    Returns:
        tuple: (
            merged tasks dictionary,
            tasks to be created in remote,
            tasks to be updated in remote,
            tasks to be created in local,
            tasks to be updated in local
        )
    """
    merged_tasks = {}
    to_create_remote = {}
    to_update_remote = {}
    to_create_local = {}
    to_update_local = {}

    # Step 1: Identify all unique task IDs across both sources
    all_task_ids = set(local_tasks.keys()) | set(remote_tasks.keys())

    for task_id in all_task_ids:
        local_task = local_tasks.get(task_id)
        remote_task = remote_tasks.get(task_id)

        # Case 1: Task exists only locally - add to remote
        if local_task and not remote_task:
            merged_tasks[task_id] = local_task
            to_create_remote[task_id] = local_task

        # Case 2: Task exists only in remote - add to local
        elif not local_task and remote_task:
            merged_tasks[task_id] = remote_task
            to_create_local[task_id] = remote_task

        # Case 3: Task exists in both - resolve conflicts
        else:
            merged_task, should_update_local, should_update_remote = resolve_task_conflict(
                local_task, remote_task
            )

            merged_tasks[task_id] = merged_task

            if should_update_local:
                to_update_local[task_id] = merged_task

            if should_update_remote:
                to_update_remote[task_id] = merged_task

    return (
        merged_tasks,
        to_create_remote,
        to_update_remote,
        to_create_local,
        to_update_local
    )

def resolve_task_conflict(local_task, remote_task):
    """
    Resolve conflicts between two versions of the same task.

    Returns:
        tuple: (
            merged task,
            should_update_local (bool),
            should_update_remote (bool)
        )
    """
    # Make a copy of the local task to use as our base
    merged_task = copy.deepcopy(local_task)

    # Track if we need to update either source
    should_update_local = False
    should_update_remote = False

    # Most recent update wins for most fields
    if remote_task.updated_at > local_task.updated_at:
        # Remote task is newer, update local fields
        merged_task.title = remote_task.title
        merged_task.description = remote_task.description
        merged_task.priority = remote_task.priority
        merged_task.due_date = remote_task.due_date
        should_update_local = True
    else:
        # Local task is newer or same age, update remote fields
        should_update_remote = True

    # Special handling for completed status - completed wins over not completed
    if remote_task.status == TaskStatus.DONE and local_task.status != TaskStatus.DONE:
        merged_task.status = TaskStatus.DONE
        merged_task.completed_at = remote_task.completed_at
        should_update_local = True
    elif local_task.status == TaskStatus.DONE and remote_task.status != TaskStatus.DONE:
        # Keep local status (already in merged_task)
        should_update_remote = True
    elif remote_task.status != local_task.status:
        # Different non-completed status - most recent wins
        if remote_task.updated_at > local_task.updated_at:
            merged_task.status = remote_task.status
            should_update_local = True
        else:
            # Keep local status (already in merged_task)
            should_update_remote = True

    # Merge tags from both sources (union)
    all_tags = set(local_task.tags) | set(remote_task.tags)
    merged_task.tags = list(all_tags)

    # If tags changed in either source, update both
    if set(merged_task.tags) != set(local_task.tags):
        should_update_local = True
    if set(merged_task.tags) != set(remote_task.tags):
        should_update_remote = True

    # Update the timestamp to latest
    merged_task.updated_at = max(local_task.updated_at, remote_task.updated_at)

    return merged_task, should_update_local, should_update_remote

File: use-cases/testing-001/javascript/TaskManager/task_list_merge.js
const { TaskStatus } = require("./models");

function mergeTaskLists(localTasks, remoteTasks) {
  /**
   * Merge two task lists with conflict resolution.
   *
   * Args:
   *   localTasks: Object of tasks from local source {task_id: task}
   *   remoteTasks: Object of tasks from remote source {task_id: task}
   *
   * Returns:
   *   Object with:
   *     - mergedTasks: Combined tasks
   *     - toCreateRemote: Tasks to be created in remote
   *     - toUpdateRemote: Tasks to be updated in remote
   *     - toCreateLocal: Tasks to be created in local
   *     - toUpdateLocal: Tasks to be updated in local
   */
  const mergedTasks = {};
  const toCreateRemote = {};
  const toUpdateRemote = {};
  const toCreateLocal = {};
  const toUpdateLocal = {};

  // Step 1: Identify all unique task IDs across both sources
  const allTaskIds = new Set([
    ...Object.keys(localTasks),
    ...Object.keys(remoteTasks)
  ]);

  for (const taskId of allTaskIds) {
    const localTask = localTasks[taskId];
    const remoteTask = remoteTasks[taskId];

    // Case 1: Task exists only locally - add to remote
    if (localTask && !remoteTask) {
      mergedTasks[taskId] = localTask;
      toCreateRemote[taskId] = localTask;
    }
    // Case 2: Task exists only in remote - add to local
    else if (!localTask && remoteTask) {
      mergedTasks[taskId] = remoteTask;
      toCreateLocal[taskId] = remoteTask;
    }
    // Case 3: Task exists in both - resolve conflicts
    else {
      const [mergedTask, shouldUpdateLocal, shouldUpdateRemote] =
        resolveTaskConflict(localTask, remoteTask);

      mergedTasks[taskId] = mergedTask;

      if (shouldUpdateLocal) {
        toUpdateLocal[taskId] = mergedTask;
      }

      if (shouldUpdateRemote) {
        toUpdateRemote[taskId] = mergedTask;
      }
    }
  }

  return {
    mergedTasks,
    toCreateRemote,
    toUpdateRemote,
    toCreateLocal,
    toUpdateLocal
  };
}

function resolveTaskConflict(localTask, remoteTask) {
  /**
   * Resolve conflicts between two versions of the same task.
   *
   * Returns:
   *   [mergedTask, shouldUpdateLocal, shouldUpdateRemote]
   */
  // Make a copy of the local task to use as our base
  const mergedTask = {...localTask};

  // Track if we need to update either source
  let shouldUpdateLocal = false;
  let shouldUpdateRemote = false;

  // Most recent update wins for most fields
  const localDate = new Date(localTask.updatedAt);
  const remoteDate = new Date(remoteTask.updatedAt);

  if (remoteDate > localDate) {
    // Remote task is newer, update local fields
    mergedTask.title = remoteTask.title;
    mergedTask.description = remoteTask.description;
    mergedTask.priority = remoteTask.priority;
    mergedTask.dueDate = remoteTask.dueDate;
    shouldUpdateLocal = true;
  } else {
    // Local task is newer or same age, update remote fields
    shouldUpdateRemote = true;
  }

  // Special handling for completed status - completed wins over not completed
  if (remoteTask.status === TaskStatus.DONE && localTask.status !== TaskStatus.DONE) {
    mergedTask.status = TaskStatus.DONE;
    mergedTask.completedAt = remoteTask.completedAt;
    shouldUpdateLocal = true;
  } else if (localTask.status === TaskStatus.DONE && remoteTask.status !== TaskStatus.DONE) {
    // Keep local status (already in mergedTask)
    shouldUpdateRemote = true;
  } else if (remoteTask.status !== localTask.status) {
    // Different non-completed status - most recent wins
    if (remoteDate > localDate) {
      mergedTask.status = remoteTask.status;
      shouldUpdateLocal = true;
    } else {
      // Keep local status (already in mergedTask)
      shouldUpdateRemote = true;
    }
  }

  // Merge tags from both sources (union)
  const allTags = [...new Set([...localTask.tags, ...remoteTask.tags])];
  mergedTask.tags = allTags;

  // If tags changed in either source, update both
  if (!arraysEqual(mergedTask.tags, localTask.tags)) {
    shouldUpdateLocal = true;
  }
  if (!arraysEqual(mergedTask.tags, remoteTask.tags)) {
    shouldUpdateRemote = true;
  }

  // Update the timestamp to latest
  mergedTask.updatedAt = localDate > remoteDate ? localTask.updatedAt : remoteTask.updatedAt;

  return [mergedTask, shouldUpdateLocal, shouldUpdateRemote];
}

// Helper function to compare arrays
function arraysEqual(a, b) {
  if (a.length !== b.length) return false;
  const sortedA = [...a].sort();
  const sortedB = [...b].sort();
  return sortedA.every((val, i) => val === sortedB[i]);
}

// Export functions for testing
module.exports = { mergeTaskLists, resolveTaskConflict, arraysEqual };
